File: chapters/10-database-design-and-data-modeling/java/OptimisticVersionUpdate.java
// Scenario: inventory reservation updates a row with version check to avoid lost writes
// Demonstrates: optimistic concurrency (version column) on contested order/inventory rows
// Trade-off: retry on conflict vs pessimistic SELECT FOR UPDATE (holds locks longer)

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.util.Optional;

final class OptimisticVersionUpdate {
  Optional<Long> reserveStock(Connection conn, String sku, int qty, long expectedVersion)
      throws Exception {
    String sql = """
        UPDATE inventory
        SET on_hand = on_hand - ?, version = version + 1
        WHERE sku = ? AND version = ? AND on_hand >= ?
        """;
    try (PreparedStatement ps = conn.prepareStatement(sql)) {
      ps.setInt(1, qty);
      ps.setString(2, sku);
      ps.setLong(3, expectedVersion);
      ps.setInt(4, qty);
      if (ps.executeUpdate() == 1) {
        return Optional.of(expectedVersion + 1);
      }
      return Optional.empty(); // caller reloads and retries or returns 409
    }
  }
}

File: chapters/10-database-design-and-data-modeling/go/optimistic_version_update.go
// Scenario: inventory reservation updates a row with version check to avoid lost writes
// Demonstrates: optimistic concurrency via version column in UPDATE ... WHERE
// Trade-off: explicit retry in app vs long-lived row locks under pessimistic locking

package main

import (
	"context"
	"database/sql"
	"errors"
)

var ErrVersionConflict = errors.New("inventory version conflict")

func ReserveStock(ctx context.Context, db *sql.DB, sku string, qty int, expectedVersion int64) (int64, error) {
	res, err := db.ExecContext(ctx, `
		UPDATE inventory
		SET on_hand = on_hand - $1, version = version + 1
		WHERE sku = $2 AND version = $3 AND on_hand >= $1`,
		qty, sku, expectedVersion)
	if err != nil {
		return 0, err
	}
	n, _ := res.RowsAffected()
	if n == 1 {
		return expectedVersion + 1, nil
	}
	return 0, ErrVersionConflict
}
